src/core/ext/filters/client_channel/xds/xds_bootstrap.cc

```diff
@@ -246,242 +187,81 @@ grpc_error* XdsBootstrap::ParseChannelCreds(grpc_json* json, size_t idx,
   return error;
 }
 
-grpc_error* XdsBootstrap::ParseNode(grpc_json* json) {
+grpc_error* XdsBootstrap::ParseNode(Json* json) {
   InlinedVector<grpc_error*, 1> error_list;
   node_ = MakeUnique<Node>();
-  bool seen_metadata = false;
-  bool seen_locality = false;
-  for (grpc_json* child = json->child; child != nullptr; child = child->next) {
-    if (child->key == nullptr) {
-      error_list.push_back(
-          GRPC_ERROR_CREATE_FROM_STATIC_STRING("JSON key is null"));
-    } else if (strcmp(child->key, "id") == 0) {
-      if (child->type != GRPC_JSON_STRING) {
-        error_list.push_back(GRPC_ERROR_CREATE_FROM_STATIC_STRING(
-            "\"id\" field is not a string"));
-      }
-      if (node_->id != nullptr) {
-        error_list.push_back(
-            GRPC_ERROR_CREATE_FROM_STATIC_STRING("duplicate \"id\" field"));
-      }
-      node_->id = child->value;
-    } else if (strcmp(child->key, "cluster") == 0) {
-      if (child->type != GRPC_JSON_STRING) {
-        error_list.push_back(GRPC_ERROR_CREATE_FROM_STATIC_STRING(
-            "\"cluster\" field is not a string"));
-      }
-      if (node_->cluster != nullptr) {
-        error_list.push_back(GRPC_ERROR_CREATE_FROM_STATIC_STRING(
-            "duplicate \"cluster\" field"));
-      }
-      node_->cluster = child->value;
-    } else if (strcmp(child->key, "locality") == 0) {
-      if (child->type != GRPC_JSON_OBJECT) {
-        error_list.push_back(GRPC_ERROR_CREATE_FROM_STATIC_STRING(
-            "\"locality\" field is not an object"));
-      }
-      if (seen_locality) {
-        error_list.push_back(GRPC_ERROR_CREATE_FROM_STATIC_STRING(
-            "duplicate \"locality\" field"));
-      }
-      seen_locality = true;
-      grpc_error* parse_error = ParseLocality(child);
-      if (parse_error != GRPC_ERROR_NONE) error_list.push_back(parse_error);
-    } else if (strcmp(child->key, "metadata") == 0) {
-      if (child->type != GRPC_JSON_OBJECT) {
-        error_list.push_back(GRPC_ERROR_CREATE_FROM_STATIC_STRING(
-            "\"metadata\" field is not an object"));
-      }
-      if (seen_metadata) {
-        error_list.push_back(GRPC_ERROR_CREATE_FROM_STATIC_STRING(
-            "duplicate \"metadata\" field"));
-      }
-      seen_metadata = true;
-      InlinedVector<grpc_error*, 1> parse_errors =
-          ParseMetadataStruct(child, &node_->metadata);
-      if (!parse_errors.empty()) {
-        grpc_error* parse_error = GRPC_ERROR_CREATE_FROM_VECTOR(
-            "errors parsing \"metadata\" object", &parse_errors);
-        error_list.push_back(parse_error);
-      }
+  auto it = json->mutable_object()->find("id");
+  if (it != json->mutable_object()->end()) {
+    if (it->second.type() != Json::Type::STRING) {
```

Comment:
Confirm `id` field is optional?

Reply:
Yeah, I don't see any reason it should be required.  Even if the server wants to require it, the client doesn't need to impose that requirement.

Reply:
Hmm ok. It was just a bit weird to see that a node (or anything) having an optional "identifier" field. But I guess it's possible if the picking algorithm just doesn't care which node it is.